### ISSUE TITLE
Adblock/Umatrix like functionality?

### ISSUE THREAD

**Issue by schvabodka-man** [atlas-engineer/nyxt#32]
Hello. Are there plans on implementing per domain based settings to block stuff like ads, javascript, pictures, etc like umatrix?

**Comment by jmercouris** [comment]
Hi, thank you for opening your issue! I share your sentiment, as a user, Adblock is immensely important to me/blocking images/js etc

So to answer your question, yes there are definitely plans for this, I think probably in version 0.07 or 0.08! I'll keep you updated!

-John

**Comment by stellarstriker** [comment]
+1

**Comment by jmercouris** [comment]
Hey everyone, just to provide a quick update on this.  I am still working hard on the new version which will finally enable this functionality. I anticipate a release within a couple of weeks. Thank you for your patience!

-John

**Comment by impaktor** [comment]
Will it be some native solution, or will it rely on existing addons,  like [uBlock origin](https://github.com/gorhill/uBlock) or Adblock plus? The former being [much preferred](https://medium.com/@trybravery/please-stop-using-adblock-but-not-why-you-think-13280e76c8e7).

**Comment by jmercouris** [comment]
Well, it turns out that my integration efforts have been in vain, because I'd have to make my own webkit port on MacOS to actually use the purported content blocking rules (only present in Safari), if any one has any other information they are privy to, or knows how to do it, please let me know.


As of now, the follow up plan is to implement advertisement blocking by updating the hosts file, not exactly the most elegant, but it works...

**Comment by jgkamat** [comment]
For simple host based blocking (at least), this can be done a bit cleaner (and without editing the system hosts file, which will probably need root) in qtwebkit by [setting a network access manager](http://doc.qt.io/archives/qt-5.5/qwebpage.html#setNetworkAccessManager) which you [define yourself](https://doc.qt.io/qt-5/qnetworkaccessmanager.html) to block whatever domains you want. This the example of [how qb does this on qtwebkit](https://github.com/qutebrowser/qutebrowser/blob/3170e35b31a500a1729ee2ecafd231d55535ab57/qutebrowser/browser/webkit/network/networkmanager.py#L355-L410). I'm pretty sure there should be something similar in webkitgtk, as [luakit](https://luakit.github.io/) has a hostblocker built in, but I'm much less familiar with that (you might be able to get some pointers by reading the relevant sections of their source!)

**Comment by jmercouris** [comment]
Hi, to everyone wanting this ad-blocking feature. I am hard at work, but I need your help to make it happen, as part of a new architecture, I did finally figure out how to actually make ad-block possible. 

If you would be able to, please support, or spread awareness about my crowdfunding campaign so that I can pay the C/C++ developer who I'll be working with, thank you for your time!

https://www.indiegogo.com/projects/next-browser-nix-support#/

**Comment by Ambrevar** [comment]
I'm on this!  Expect at least a host-based blocker + image blocker + javascript toggle from the work I'm going to put on this in the coming days :D

**Comment by Ambrevar** [comment]
I've just pushed the resource blocker.  Javascript toggle coming soon.
Not too sure about the image blocker actually, does anyone know how to deal with image in WebKitGTK?

**Comment by impaktor** [comment]
Nice! I assume it will be documented as well?

**Comment by Ambrevar** [comment]
It already is (a little):
https://github.com/atlas-engineer/next/blob/master/documents/MANUAL.org#resource-blodking-ad-blocking

**Comment by whirm** [comment]
Would it be hard to integrate this: https://github.com/brave/ad-block ?

**Comment by Ambrevar** [comment]
We can't (or don't want to) "integrate" brace/ad-block directly since we
would like to stick to a full-Lisp code base as much as possible, but
sure the current algorithm can be expanded to follow more closely what
Ublock-origin and brace/ad-block do.

It's a bit of work, but certainly doable.  Specialists in ad-blocking
are welcome to pitch in :D

**Comment by Ambrevar** [comment]
See also #464.

**Comment by whirm** [comment]
Does this issue have any priority? I'd really like this to be implemented as It's the only think preventing me from switching to Next as far as I've seen.

**Comment by Ambrevar** [comment]
It has high priority indeed, post 2.0 (which is due in hopefully just a
few of weeks).

One way to handle it would be to implement support WebExtensions, then
we would get uBlock for free.

Help is more than welcome! :)

**Comment by dudik** [comment]
Hello. I'm working on a WebKitGTK extension similar to uBlock Origin as my bachelors thesis. I wanted to ask if Nyxt supports loading [WebKitWebExtensions](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebExtension.html) written in C/C++. Based on [this cl-webkit code](https://github.com/joachifm/cl-webkit/search?q=webkit_web_context_set_web_extensions_directory) I think it should, but I want to be sure so I can potentially refer to Nyxt as a reason for why the WebKitGTK ecosystem would benefit from having such an extension and that the user base of these browsers is not negligible. Thank you.

**Comment by Ambrevar** [comment]
We were just looking for a reason to merge WebExtension support!

https://github.com/joachifm/cl-webkit/pull/41

@aartaka Any comments?

**Comment by aartaka** [comment]
It seems we found the reason to complete WebKitWebExtension support :tada:

I can look into it next week and test it with example extensions I found. Shouldn't be hard, so by the end of the next week we should have support for WebKitWebExtensions loading!

@dudik, thank you for this motivation!

**Comment by dudik** [comment]
It makes me extremely happy to see your enthusiasm. Sadly, I have just started writing my thesis and the first half is centered around theory. I'm also still in the middle of a normal semester, so I won't be able to show you anything worthwhile until February. Maybe sooner, but it's hard to say right now. I made some research and it is also in my own interest to have a WebKitGTK content blocker, so I'm positive about creating something useful.
I'm really sorry, but also very thankful for the motivation.

**Comment by aartaka** [comment]
No pressure, you're not obliged to work on it just because we found your idea to be in-line with what we do :)

In any way, good luck!

**Comment by aartaka** [comment]
@dudik, good news: we've merged the `WebKitWebExtension` support in cl-webkit! This means that Nyxt can use WebKit extensions and you can refer to Nyxt as a use-case for your thesis :)

**Comment by kskyten** [comment]
Qutebrowser recently merged an adblocker based on Brave qutebrowser/qutebrowser#5317. Perhaps worth looking at how they did it. As far as I know, it is implemented in Rust, so it would add a non-lisp part into the code base, though.

**Comment by jmercouris** [comment]
It should not be part of the core, we would like to support the actual webextensions API so firefox and other plugins can be easily used and ported

**Comment by jmercouris** [comment]
In other words, no Nyxt specific hacks should be necessary

**Comment by jmercouris** [comment]
In other words, no Nyxt specific hacks should be necessary

**Comment by jmercouris** [comment]
congratulations to the team for the achievement!

**Comment by dudik** [comment]
So after working on a pure C adblock for a month, I realized that because of performance reasons, I would have to implement it with inter-process communication in mind. That gave me the freedom to use any language I want so I ended up putting the C adblock aside and just creating a "bridge" between the Brave Rust adblocker and the C extension. [This is the result](https://github.com/dudik/blockit). At this time it's pretty barebones and the code quality is questionable, but it's working both for cosmetic and network filtering so I felt it was the right time to make it public. While writing this comment I realized that I'm not freeing some memory in the C extension. This will get fixed tommorow when I wake up as I'm pretty sleepy. Anyways, in the meantime I would like to hear if it works on your platform and how well. Thanks!

**Comment by Ambrevar** [comment]
Thank you for much for sharing your work!

I just had a look at the code: so it looks like if we want to integrate
this in Nyxt, we just need to write FFI bindings to `void
webkit_web_extension_initialize(WebKitWebExtension *extension)`, is this correct?

**Comment by aartaka** [comment]
No, we don't :)

This function is an entry point for all WebKit extensions, so the only thing we need is to compile the extension, put it where Nyxt can find it (see [WebKitWebExtension documentation](https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebExtension.html)), run adblock server, and enjoy the extension being enabled in Nyxt on the next startup.

**Comment by Ambrevar** [comment]
You mean we already have this binding?

**Comment by aartaka** [comment]
Yes and no 🙃 

We don't have it as a CL-CFFI binding, but we don't need to have it because WebKit manages it's extensions automatically. The only change to Nyxt that we need to do is to set the extensions directory. This is trivial, I'll open the PR, most probably in a couple of hours.

**Comment by whirm** [comment]
@aartaka are you still planning on open that PR? Cheers!

**Comment by aartaka** [comment]
> @aartaka are you still planning on open that PR? Cheers!

Actually, it's [long merged](https://github.com/atlas-engineer/nyxt/pull/1189) :D 

The directory is .config/nyxt/extensions/. However, it's likely to change soon (as per https://github.com/atlas-engineer/nyxt/pull/1505), so keep an eye on the `gtk-extensions-path` of your Nyxt buffers :)

**Comment by whirm** [comment]
Amazing, I'll give it a go ASAP. Thanks a lot!

**Comment by Ambrevar** [comment]
Of course, and some exist already! See https://github.com/atlas-engineer/nyxt/blob/master/documents/EXTENSIONS.org.

As for ad-blocking, the problem is mostly with maintenance; it's a huge endeavor.  We would be better off reusing a well maintained ad-blocker than rolling out our own.

**Comment by arne-cl** [comment]
> The directory is .config/nyxt/extensions/. However, it's likely to change soon (as per #1505), so keep an eye on the `gtk-extensions-path` of your Nyxt buffers :)

@aartaka Could you please give an example of how to load a WebKitWebExtension in Nyxt?

**Comment by aartaka** [comment]
> @aartaka Could you please give an example of how to load a WebKitWebExtension in Nyxt?

Sure! I assume you are aware of the fact that WebKitWebExtensions are `.so` C libraries. The steps are:
- Get an extension from somewhere, either as binary or as the source code.
- (if it's not yet a binary) compile it according to instructions in the original repo.
- Put the resulting `.so` in `.local/shane/nyxt/extensions`.
- Start Nyxt.
- The extension should be automatically loaded into every buffer (except for the internal ones, like the initial help buffer).

One concern of mine, that we currently discussed in #1781, is that it can actually not load the extension due to WebKit being picky about the directories. To trespass the restriction, you can do `export WEBKIT_FORCE_SANDBOX=0` (or add `(setf (uiop:getenv "WEBKIT_FORCE_SANDBOX") "0")` to your config), until we come up with a better solution. Beware: this can disable extension sandboxing, so be careful about what you install!

**Comment by crocket** [comment]
> I'd say WebExtensions, as those can hijack JS requests and modify the page opened in the browser, while any external adblocking library is usually restricted to host filtering.

Perhaps, that's why qutebrowser can't do cosmetic filtering, yet?

**Comment by glyh** [comment]
I think it'll be better if we support web-extension apis, and then a lot of functionality can be done at the plugin side rather than embrace everything in the browser itself, which can ease the pain of maintaining a big project.

**Comment by aartaka** [comment]
Now that #1762 is merged, there's a solid base to build uBlock support on top of. Here's a list of things to do for uBlock to work (copy-pasted from the linked pull request):
  - Methods (alphabetically, with unimplemented on top):
    - [ ] browser.browserAction.setPopup -- easy
    - [ ] browser.commands.getAll -- easy
    - [ ] browser.commands.reset -- easy
    - [ ] browser.commands.update -- easy
    - [ ] browser.i18n.getMessage -- easy?
    - [ ] browser.runtime.connect
    - [ ] browser.runtime.reload
    - [ ] browser.tabs.move -- easy
    - [ ] browser.webRequest.filterResponseData
    - [ ] browser.webRequest.handlerBehaviorChanged
    - [X] browser.runtime.getBrowserInfo
    - [X] browser.runtime.getManifest
    - [X] browser.runtime.getURL
    - [X] browser.tabs.create
  - Variables and types (will be mostly easy to do):
    - [ ] browser.runtime.lastError
    - [ ] browser.windows.WINDOW_ID_NONE -- easy
    - [ ] browser.webRequest.ResourceType -- easy
  - Events (alphabetically, with unimplemented on top, hardest to do as they require modifications all over Nyxt sources):
    - [ ] browser.browserAction.onClicked -- easy
    - [ ] browser.commands.onCommand -- easy
    - [ ] browser.runtime.onConnect
    - [ ] browser.runtime.onUpdateAvailable
    - [ ] browser.tabs.onActivated
    - [ ] browser.tabs.onCreated -- easy
    - [ ] browser.tabs.onRemoved -- easy
    - [ ] browser.tabs.onUpdated -- easy
    - [ ] browser.webNavigation.onCommitted
    - [ ] browser.webNavigation.onCreatedNavigationTarget
    - [ ] browser.webRequest.onBeforeRequest
    - [ ] browser.webRequest.onHeadersReceived
    - [ ] browser.windows.onFocusChanged -- easy
  - Non-mandatory but cool APIs because uBlock uses them to provide more features:
    - [ ] browser.dns
    - [ ] browser.privacy
    - [ ] browser.menus

**Comment by crocket** [comment]
While it's good to have adblock in a web browser, Rob Braxman recommends blocking ads in your personal recursive DNS server because uBlock Origin makes browser fingerprint more unique.

**Comment by crocket** [comment]
Does nyxt's host blocker automatically update its block list?

**Comment by aartaka** [comment]
> Does nyxt's host blocker automatically update its block list?

Yes, it does. The update frequency (in seconds) can be configured using `nyxt/blocker-mode:update-interval` slot of `nyxt/blocker-mode:blocker-mode`.

**Comment by aartaka** [comment]
Oh, wait, I'm wrong here :sweat_smile: It's `nyxt/blocker-mode:update-interval` of every hostlist that you need to manually alter, i.e.
``` lisp
(define-configuration nyxt/blocker-mode:blocker-mode
  ((nyxt/blocker-mode:hostlists
    (mapcar (lambda (h)
              ;; Update every 2 minutes
              (setf (nyxt/blocker-mode:update-interval) 120))
            %slot-default%))))
```

**Comment by neophrema** [comment]
I'm sorry to ask explicitly, are there plans to make something like **uMatrix** work? I prefer it to just allowing scripts as whole or not. I could imagine that the integration into nyxt would be rather neat, if possible. The Privacy issues are the only thing holding me back from using nyxt, and I would be super happy to use it.

**Comment by crocket** [comment]
uMatrix is unmaintained, but it still works.
Host blocker already takes care of most of privacy issues.

**Comment by neophrema** [comment]
I need to frankly disagree @crocket
It might be much more convenient to follow a "enable all or nothing" setting, but as a long time user of uMatrix I'm very positive about that its a necessary feature regarding privacy. I know this sounds demanding, especially because I can't get involved into making it true for nyxt, and I feel sorry for that.

It appears to me that its rather the majority of sites don't need half of external site loads they establish, to work. This is in fact a huge tracking possibility. What do you think?

**Comment by crocket** [comment]
Websites track you via browser fingerprint. The more add-ons you add to a web browser, the more unique your browser fingerprint becomes.

If you want privacy, you need to hide your IP address and compartmentalize different web activities to different web browsers.
Specifically, you need a separate web browser for google accounts and facebook accounts because google and facebook have global infrastructure to store browser fingerprints and track users on various websites that use functionalities from google and facebook.

uMatrix can help, but uBlock Origin can also block third party domains by default.

**Comment by neophrema** [comment]
Thank you for your explanation!
- The tradeoff seems to be: uMatrix blocks the potential connection to 50% of 3rd party websites. These don't get to fingerprint me, don't get metadata **but** the firstparty websites counts me to the pool of uMatrix users?
- If I'm for example streaming some media I need one 3rd party website, e.g. akamaihd for www.zdf.de, but I don't want to enable other 3rd party websites. Isn't it handy to block requests more granular?
- Right now I'm using: https://github.com/arkenfox/user.js/  --  If I log into Github it registers my machine as a Windows machine, im on Linux. It appears like the spoofing is working. Is something like that also thinkable under nyxt? I really, **really** fear what facebook and google are doing to our societies, and it's crucial for us to shield us and others. Filterbubbles are blowing up our systems, it appears to me.
- I'm also using a Pi-hole for my own network, but this ofc. doesn't work while being not at home.

![20220123_20h05m35s_grim](https://user-images.githubusercontent.com/48735216/150691925-e7102be7-4922-4d1a-b45d-f24de12997ee.png)

Thank you, I didn't want to cause troubles, I was just concerned, wanted to have a peek into what direction _Nyxt_ might be heading in these regards!

**Comment by crocket** [comment]
I think uBlock Origin can block with granularity if you know how to write rules.

**Comment by Svadkos** [comment]
There's UI for that for a very long time already https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-quick-guide so umatrix is redundant, if you're using ublock.

**Comment by rEnr3n** [comment]
As far as I can see, ublock origin can only block images, scripts and frames. Umatrix can also block these and more. It's shown in the picture provided by @neophrema. Umatrix is not exactly redundant, it is complementary (or supplementary). Both of these tools are made by the same author. Privacy is not the only thing umatrix addresses. I mostly use ublock origin for list-based blocking and umatrix for making webpages less fat.

**Comment by Svadkos** [comment]
Yeah and the same author has stopped Umatrix development 2 years ago.

I've used both extensively and now i don't see much point in it. Cookies aren't such a big deal with first-party isolation, and there are plenty of tools to manage them. Pictures and media can be blocked. XHR are blocked with scripts. Most of "Other" can be blocked by blocking scripts too, and there's an option to block csp reports.

There's no point to have these granular options when you can block first or third party scripts altogether. Otherwise it just gives you an illusion of control.

**Comment by rEnr3n** [comment]
So I tried to replicate my rules in umatrix to ubo. My ruleset is deny by default (except css/images). In ubo, I blocked 3rd-party scripts and frames. I noticed on this [link](https://twitter.com/Bulgariofficial) that several xhr goes through. Videos play through xhr. So blocking 3rd-party scripts is not enough. I could block 3rd-party here but then if I were to unblock, all the elements under that domain will be unblocked (images, scripts, xhr, etc). That's not what I wanted.

While ubo works for "most" people. I am one of those people that don't fall into that category. I could have settled for ubo alone if it was enough for me. But not everyone has the same preferences.

**Comment by Svadkos** [comment]
Add this rule to "My rules" tab in ubo settings.
`twitter.com https://video.twimg.com/ext_tw_video/ xmlhttprequest block`
You can create it from logger window like that
![image](https://user-images.githubusercontent.com/9449246/151960115-3fcb0a15-97a2-438e-b804-aba1fd7bc74a.png)

Also you can create your own filtering rules with the same syntax used by blacklists and there's UI for that too.

**Comment by rEnr3n** [comment]
I know I can do that but it doesn't align with my preference which is to deny by default. Twitter is not the only entity on the internet that can do that. I will have to maintain both a whitelist and a blacklist. That looks to be more exhausting than my current setup.

**Comment by Svadkos** [comment]
If you want to block everywhere add this on My filters tab
```
*$3p,xhr
```
And these are exceptions for scripts to work
```
! exceptions
@@||abs.twimg.com/*js|$domain=twitter.com
@@||edgecastcdn.net$domain=twitter.com
```
Also, you can combine filters and rules, where the latter takes precedence, so exceptions can be set there too, but filters are much more flexible.

You don't have to write filters by hand, because there's a UI for that in logger, but here's a good overview of filtering syntax https://adblockplus.org/filter-cheatsheet and ubo-specific options there https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-rule-syntax

But then again in case of twitter I would rather use a nitter instance.

**Comment by aadcg** [comment]
To summarize:

For a primitive adblocker that works for any port see `blocker-mode`.

For the WebKitGTK port, you can give [blockit](https://github.com/dudik/blockit) a try. For setup instructions see https://github.com/atlas-engineer/nyxt/pull/3562. Supporting WebExtensions on WebKitGTK requires lots of work and is out of scope right now.

For the Electron port, it should be easier to run web extensions. Also, there's https://github.com/ghostery/adblocker.

--------

Regardless, we need to focus on more pressing issues.
